lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,427 @@
+    """Nelder Mead Optimizer
+
+    The Nelder-Mead optimizer is an algorithm used for unconstrained
+    optimization in multidimensional spaces. Unlike some other
+    optimization methods, it does not take into account any bounds or
+    constraints on the variables. The algorithm is based on the Simplex
+    method and is particularly useful when the derivatives
+    (first and second)  of the objective function are unknown or
+    unreliable.
+
+    Although the Nelder-Mead algorithm is generally robust and widely
+    applicable, it has some limitations. In cases where the derivatives
+    can be accurately computed, alternative algorithms that utilize
+    this derivative information may offer better performance. These
+    methods are often preferred due to their ability to handle a wider
+    range of scenarios and their tendency to converge to more optimal
+    solutions. It is worth noting that the Nelder-Mead technique is a
+    heuristic search approach, which means that it can sometimes
+    converge to non-stationary points or suboptimal solutions.
+
+    This implementation is based heavily scipy's optimize.minimize. See
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
+
+    """
+
+    model: QuantumModel
+
+    def __init__(
+        self,
+        model: QuantumModel,
+        hyperparams: dict[str, float],
+        loss_fn: Callable[[Any, Any], float],
+        bounds: ArrayLike | None = None,
+    ) -> None:
+        """Initialise the Nelder-Mead optimizer.
+
+        The hyperparameters may contain the following key value pairs:
+
+        - `adaptive`: Adjust the algorithm's parameters based on the
+                dimensionality of the problem. This adaptation is
+                particularly helpful when minimizing functions in
+                high-dimensional spaces, bool.
+
+        - `maxfev`: Maximum number of function evaluations allowed.
+                Default is 1000. int.
+
+        - `initial_simplex`: If provided, the `initial simplex`
+                replaces the initial model weights. Each row
+                initial_simplex[i, :] should contain the coordinates
+                of the ith vertex among the N+1 vertices in the
+                simplex, where N represents the dimension,
+                ArrayLike (N+1, N), float.
+
+        - `xatol`: The acceptable level of absolute error in the
+                optimal model weights (optimal solution) between
+                iterations that indicates convergence, float.
+
+        - `fatol`: The acceptable level of absolute error in the
+                loss value between iterations that indicates convergence,
+                float.
+        }
+
+        Parameters
+        ----------
+        model : :py:class:`.QuantumModel`
+            A lambeq quantum model.
+        hyperparams : dict of str to float.
+            A dictionary containing the models hyperparameters.
+        loss_fn : Callable
```

Comment:
Is it possible to specify the input and output types of this? If it's not simple, then just leave it as it is